mm2src/coins/z_coin.rs

```diff
@@ -667,6 +669,33 @@ impl ZCoin {
             paging_options: request.paging_options,
         })
     }
+
+    /// Validates dex fee output or burn output
+    /// Returns true if the output valid or error if not valid. Returns false if could not decrypt output (some other output)
+    fn validate_dex_fee_output(
+        &self,
+        shielded_out: &OutputDescription,
+        ovk: &OutgoingViewingKey,
+        expected_address: &PaymentAddress,
+        block_height: BlockHeight,
+        amount_sat: u64,
+        expected_memo: &MemoBytes,
+    ) -> Result<bool, String> {
+        if let Some((note, address, memo)) =
+            try_sapling_output_recovery(self.consensus_params_ref(), block_height, ovk, shielded_out)
+        {
+            if &address == expected_address {
+                if note.value != amount_sat {
+                    return Err(format!("invalid amount {}, expected {}", note.value, amount_sat));
+                }
+                if &memo != expected_memo {
+                    return Err(format!("invalid memo {:?}, expected {:?}", memo, expected_memo));
+                }
+                return Ok(true);
+            }
+        }
+        Ok(false)
```

Comment:
we can use sugar and avoid nesting `if` blocks

```rust
        let Some((note, address, memo)) =
            try_sapling_output_recovery(self.consensus_params_ref(), block_height, ovk, shielded_out)
        else {
            return Ok(false);
        };

        if &address != expected_address {
            return Ok(false);
        }

        if note.value != amount_sat {
            return Err(format!("Invalid amount: got {}, expected {}", note.value, amount_sat));
        }

        if &memo != expected_memo {
            return Err(format!("Invalid memo: got {:?}, expected {:?}", memo, expected_memo));
        }

        Ok(true)
```

Reply:
fixed thanks